### PR TITLE
Speeding up Seeding

### PR DESCRIPTION
**Pull request by VinInn**
Major speedup of seeding (33%)
Besides usual numerical performance improvements obtained reducing accuracy, the major contribution to the speed up are:
1) remove MultipleScattering (MS) contribution from GlobalRegion but for PixelPairs
2) use average MS for doublet-building
3) do not consider a forward layer if the doublet points in the opposite direction (sic)
this last improvement shows that some more pre-filtering at doublet-level may produce even more speedup....

minor regression observed in ttbar PU15
http://innocent.home.cern.ch/innocent/RelVal/pu15_80_msSpeed21/
(out of the box does not work due to a bug whose correction is still pending..)
my personal opinion is that on the tracking side it does not look worse than before

took the opportunity for some cleanup

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by slava77** [comment]
@VinInn 
please make sure this all still works fine for at least PU35 .
This is an average expected PU  in 2016 and 80X is the release for 2016.

**Comment by cvuosalo** [comment]
@cmsbuild please test
New IB may have fix

**Comment by cvuosalo** [comment]
@smuzaffar, @degano: How can we get Jenkins to use the new, fixed IB, CMSSW_8_0_X_2015-12-10-1100?

**Comment by smuzaffar** [comment]
@cvuosalo , jenkins will start using it once it is installed in AFS and base line for comparison is done. I see that release is being installed, so I think in 30 mins it should be available for jenkins to use it

**Comment by smuzaffar** [comment]
please test
11h00 ib avaiable now

**Comment by VinInn** [comment]
@smuzaffar 
do not understand is clang or gcc?
clang++ -c -DGNU_GCC ....
I've protected my code with 
if defined(__GNUC__) 
to be used with gcc only (no clang, no icc, no whatever other compiler CMS decide to try....)

how to protect code for gcc only?

**Comment by VinInn** [comment]
comparison is line with my expectations

**Comment by smuzaffar** [comment]
@VinInn , looks like clang also defines __GNUC__
```
> echo | clang++ -dM -E - | grep GNUC
#define __GNUC_MINOR__ 2
#define __GNUC_PATCHLEVEL__ 1
#define __GNUC_STDC_INLINE__ 1
#define __GNUC__ 4
```
may be use something like
```
#ifdef __GNUC__
    #ifndef __clang__
    #endif
#endif
```

**Comment by davidlt** [comment]
Clang always defined GNUC as 4.2.1 for compatibility reasons, but you should ignore.
Thus for GCC you need `#if defined(__GNUC__) && !defined(__clang__)`

**Comment by VinInn** [comment]
@smuzaffar ,
 what about icc?
maybe I change to 
```
!defined(__INTEL_COMPILER) || !defined(__clang__)
```

**Comment by davidlt** [comment]
@VinInn IIRC, ICC reports the same version as GCC installed on the system. I.e. 4.9.3 at this point.

**Comment by VinInn** [comment]
Notice that @davidlt (I suppose) defined convenient vectors for clang as well in ExtVec.
used that and protected for icc

**Comment by slava77** [comment]
@VinInn I haven't looked at the code yet partly because from the comments in the thread it sounded like more updates are coming (and you haven't triggered the tests either after the last update).
Please clarify if changes are coming or if it's essentially ready for review.

**Comment by VinInn** [comment]
This PR is complete.
More may come in other areas of tracking
Have no triggered the test to avoid overloading of the system as the fix affects only clang

File: RecoTracker/TkDetLayers/test/TkDetLayersAnalyzer.cc [hunk 1]
```diff
@@ -116,8 +117,9 @@ TkDetLayersAnalyzer::analyze( const Event& iEvent, const EventSetup& iSetup )
     << " And Contains  Daughters: "<< (*pDD).components().size() ;   
 
   ESHandle<TrackerTopology> tTopo;
-  iSetup.get<IdealGeometryRecord>().get(tTopo);
+  iSetup.get<TrackerTopologyRcd>().get(tTopo);
 
+  /*
```
Comment (slava77) [inline RecoTracker/TkDetLayers/test/TkDetLayersAnalyzer.cc]: why not remove this code completely?
If it's needed, better put a define/ifdef  and keep it alive.

File: RecoTracker/TkHitPairs/src/InnerDeltaPhi.cc [hunk 2]
```diff
@@ -158,26 +255,32 @@ PixelRecoRange<float> InnerDeltaPhi::phiRange(const Point2D& hitXY,float hitZ,fl
...
     // add multiple scattering correction
+
+    /*
     PixelRecoPointRZ zero(0., theVtxZ);
     PixelRecoPointRZ point(hitXY.mag(), hitZ);
-    auto scatt = 3.f*sigma(thePtMin,zero, point, ol) / rLayer; 
-   
-    margin += scatt ;
+    auto scatt = 3.f*sigma(thePtMin,zero, point, ol); 
+    */
```
Comment (slava77) [inline RecoTracker/TkHitPairs/src/InnerDeltaPhi.cc]: will this commented out code be needed?

**Comment by slava77** [comment]
@cmsbuild please test
to see that clang still works.
From visual inspection, I don't see any issues other than commented out code, which could be removed instead.

File: Geometry/TrackerGeometryBuilder/test/TrackerDigiGeometryAnalyzer.cc [hunk 3]
```diff
@@ -51,6 +51,10 @@
 // class decleration
 //
 
+
+// #define PRINT(X) edm::LogInfo(X)
+#define PRINT(X) std::cout << X << ':'
```
Comment (Dr15Jones) [inline Geometry/TrackerGeometryBuilder/test/TrackerDigiGeometryAnalyzer.cc]: cout should not be used

Comment (slava77) [inline Geometry/TrackerGeometryBuilder/test/TrackerDigiGeometryAnalyzer.cc]: it's a /test file. More or less anything goes in this area.
(we are not even running static analyzer on /test subdirectories)

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/12749	a046ee2
- code changes are in line with the description
    - some cleanup of commented out code is desirable
- jenkins tests pass and comparisons with baseline show a number of minor differences
    - clang version compiles
- tested locally with a larger number of samples in CMSSW_8_0_0_pre3
    - there is a reduction of about 0.3% in the number of generalTracks in dijet and in ttbar with PU events (apparently in other samples as well in a similar range); there is no visible decrease in efficiency
    - the number of seeds in the outer iterations goes down: e.g. in 25202 workflow by about 6% in tobtec. The decrease is apparently mostly at lower pt.
- technical performance looking at 25202 (ttbar PU35) shows decrease in the seeding time as expected, in addition conversionTrackCandidates time is down by almost a factor of 2. The modules with the largest time difference are:
```
   -0.695097      -0.36%       541.80 ms/ev ->       262.33 ms/ev conversionTrackCandidates
   -0.694661      -0.11%       163.29 ms/ev ->        79.10 ms/ev pixelPairElectronSeeds
   -0.364809      -0.13%       326.31 ms/ev ->       225.63 ms/ev detachedTripletStepSeeds
   -0.271499      -0.09%       310.58 ms/ev ->       236.34 ms/ev pixelPairStepSeeds
   -0.262699      -0.01%        50.06 ms/ev ->        38.44 ms/ev mixedTripletStepSeedsB
   -0.211288      -0.02%        75.55 ms/ev ->        61.11 ms/ev initialStepSeedsPreSplitting
   -0.208349      -0.02%        74.88 ms/ev ->        60.75 ms/ev initialStepSeeds
   -0.192981      -0.03%       132.91 ms/ev ->       109.52 ms/ev lowPtTripletStepSeeds
   -0.137819      -0.01%        72.32 ms/ev ->        63.00 ms/ev tobTecStepSeedsPair
   -0.111174      -0.10%       706.29 ms/ev ->       631.90 ms/ev pixelLessStepSeeds
```
Overall, about 5% reduction in reconstruction time is observed.

File: RecoTracker/TkDetLayers/test/TkDetLayersAnalyzer.cc (hunk 1 above)
Comment (VinInn) [inline RecoTracker/TkDetLayers/test/TkDetLayersAnalyzer.cc]: it is a test, geometry is in fux (phase2 integration), let's keep it as an example...

File: RecoTracker/TkHitPairs/src/InnerDeltaPhi.cc (hunk 2 above)
Comment (VinInn) [inline RecoTracker/TkHitPairs/src/InnerDeltaPhi.cc]: it is the old version. I tend to keep previous versions around for a while to compare with the new version if issues arise 
can be protected with LogDebug or similar. (you may have noticed that I have deleted "old version" in few other placed)

**Comment by davidlt** [comment]
@VinInn this brakes ICC builds:
```
RecoTracker/TkHitPairs/src/InnerDeltaPhi.cc(234): error: identifier "cropped_asin" is undefined
```

`cropped_asin` is only available if `VI_DEBUG` macro is defined.

Could you fix it?

**Comment by VinInn** [comment]
@davidlt #12904 should fix it

**Pull request merged**